using-kubecost/admission-controller.md

```diff
@@ -0,0 +1,37 @@
+# Admission Controller
+
+{% hint style="info" %}
+The Admission Controller is a beta feature. Please read the documentation carefully.
+{% endhint %}
+
+Kubecost's Admission Controller is a tool which leverages Kubecost's Predict API to display expected future costs for changes made to CPU and RAM in your workloads. This can be used to anticipate future spend before resources have been allocated.
+
+The Admission Controller exists in Kubecost's cost analyzer pod and uses a `ValidatingAdmissionWebhook` to send back cost data on every update you make to deployments in your Kubernetes clusters. The Predict API then determines the differences in cost, and provides that information in a convenient table in your terminal.
+
+## Installing the Admission Controller
+
+### Prerequisites
+
+Before installing the Admission Controller, make sure you have [`kubectl`](https://kubernetes.io/docs/reference/kubectl/) installed.
+
+### Installation
+
+The following command will enable the Admission Controller in your existing Kubecost application:
+
+```
+helm upgrade --install kubecost \
+  --repo https://kubecost.github.io/cost-analyzer/ cost-analyzer \
+  --namespace kubecost --create-namespace \
+  --set kubecostAdmissionController.enabled=true \
+git clone https://github.com/kubecost/cost-analyzer-helm-chart.git \
+cd cost-analyzer-helm-chart/cost-analyzer/scripts \
+./create-admission-controller-tls.sh <namespace-to-install-kubecost-in>
+```
```

Comment:
Although I don't think the admission controller needs to be in the same namespace as Kubecost, let's just make it that way in this doc for simplicity. Is that ok?

Here's my suggestion:

```
# Setup TLS key/certificate for the Admission controller
$ git clone https://github.com/kubecost/cost-analyzer-helm-chart.git
$ cd cost-analyzer-helm-chart/cost-analyzer
$ kubectl create namespace kubecost
$ ./scripts/create-admission-controller-tls.sh kubecost

# Install Kubecost with AdmissionController enabled
$ helm upgrade --install kubecost \
  --repo https://kubecost.github.io/cost-analyzer/ cost-analyzer \
  --namespace kubecost --create-namespace \
  --set kubecostAdmissionController.enabled=true \
  -f values.yaml
```

Reply:
Asking users to clone an entire git repo for purposes of executing a single bash script is way overkill. This can be accomplished much more simply:

```sh
curl -s https://raw.githubusercontent.com/kubecost/cost-analyzer-helm-chart/develop/cost-analyzer/scripts/create-admission-controller-tls.sh | bash
```

Note that the script defaults to a param of `kubecost` if one is not supplied, therefore no arguments are required if the default Namespace is selected. This script is predicated on the Namespace preexisting, so if used as the first step then the user will need to be instructed on how to create a new Namespace.

Reply:
Also, if using this approach the `-f values.yaml` flag will need to be omitted as this assumes the values file is available locally which, if the repo hasn't been cloned locally, it won't be. Since the `--set` directive is used, assuming no other value overrides are provided this isn't needed.

Reply:
@chipzoller Agree with all of the above! Currently the `create-admission-controller-tls.sh` attempts to find and replace the string `${CA_BUNDLE}` in the `values.yaml` file. Hence why these commands clone the repo and include `-f values.yaml`

If we change the way the script works, we can in turn make this documentation simpler? Thoughts on best practices here Chip?

Reply:
I clearly didn't pay attention, sorry about that. I'll make a slight improvement to the script so it'll optionally just print out the encoded CA to use, but otherwise just disregard my comments here.